vault/identity_store_oidc_provider.go

```diff
@@ -1219,3 +1160,657 @@ func (i *IdentityStore) pathOIDCProviderExistenceCheck(ctx context.Context, req
+	isMember, err := i.entityHasAssignment(ctx, req.Storage, entity, client.Assignments)
+	if err != nil {
+		return authResponse("", state, ErrAuthServerError, err.Error())
+	}
+	if !isMember {
+		return authResponse("", state, ErrAuthAccessDenied, "identity entity not authorized by client assignment")
+	}
+
+	// Create the auth code cache entry
+	authCodeEntry := &authCodeCacheEntry{
+		clientID:    clientID,
+		entityID:    entity.GetID(),
+		redirectURI: redirectURI,
+		nonce:       nonce,
+		scopes:      scopes,
+	}
+
+	// Validate the optional max_age parameter to check if an active re-authentication
+	// of the user should occur. Re-authentication will be requested if max_age=0 or the
+	// last time the token actively authenticated exceeds the given max_age requirement.
+	// Returning ErrAuthMaxAgeReAuthenticate will enforce the user to re-authenticate via
+	// the user agent.
+	if maxAgeRaw, ok := d.GetOk("max_age"); ok {
+		maxAge := maxAgeRaw.(int)
+		if maxAge < 0 {
+			return authResponse("", state, ErrAuthInvalidRequest, "max_age must be greater than zero")
+		}
+		if maxAge == 0 {
+			// TODO: solve for the potential UI loop here or make max_age=0 invalid
+			return authResponse("", state, ErrAuthMaxAgeReAuthenticate, "active re-authentication is required by max_age")
+		}
+
+		// Look up the token associated with the request
+		te, err := i.tokenStorer.LookupToken(ctx, req.ClientToken)
+		if err != nil {
+			return authResponse("", state, ErrAuthServerError, err.Error())
+		}
+		if te == nil {
+			return authResponse("", state, ErrAuthAccessDenied, "token associated with request not found")
+		}
+
+		// Check if the token creation time violates the max age requirement
+		now := time.Now().UTC()
+		lastAuthTime := time.Unix(te.CreationTime, 0).UTC()
+		secondsSince := int(now.Sub(lastAuthTime).Seconds())
+		if secondsSince > maxAge {
+			return authResponse("", state, ErrAuthMaxAgeReAuthenticate, "active re-authentication is required by max_age")
+		}
+
+		// Set the auth time to use for the auth_time claim in the token exchange
+		authCodeEntry.authTime = lastAuthTime
+	}
+
+	// Generate the authorization code
+	code, err := base62.Random(32)
+	if err != nil {
+		return authResponse("", state, ErrAuthServerError, err.Error())
+	}
+
+	// Cache the authorization code for a subsequent token exchange
+	if err := i.oidcAuthCodeCache.SetDefault(ns, code, authCodeEntry); err != nil {
+		return authResponse("", state, ErrAuthServerError, err.Error())
+	}
+
+	return authResponse(code, state, "", "")
+}
+
+// authResponse returns the OIDC Authentication Response. An error response is
+// returned if the given error code is non-empty. For details, see spec at
+//   - https://openid.net/specs/openid-connect-core-1_0.html#AuthResponse
+//   - https://openid.net/specs/openid-connect-core-1_0.html#AuthError
+func authResponse(code, state, errorCode, errorDescription string) (*logical.Response, error) {
```

Comment:
@calvn, @fairclothjm - We talked last week about changing the format of a successful authorize response from the standard `logical.Response` (which included Vault-y things like `lease_id`, `renewable`, etc.) to a JSON format similar to what was already done for error responses. The idea was to closer align with the spec. I checked what this looks like in the audit log and think it's okay to proceed with that approach.

This is the response portion of an audit log for a successful authorize response using this approach:
```
"response": {
    "mount_type": "identity",
    "data": {
      "http_content_type": "hmac-sha256:5986a3f279a6e8d417d805f0e9be432afca5fe2dd5a9cb91a6cf2cf761ecaeb1",
      "http_raw_body": "hmac-sha256:f20c15bb779adc8ee97d8006c3672e3f8c0fcebbc0b5201da8e0d45922bd3721",
      "http_status_code": 200
    }
}
```

Let me know your thoughts!